File: src/main/java/Node.java
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Node {
    private final String name;
    private final String url;
    private boolean dirty;
    private boolean cleaned;
    private List<Node> parents;
    private Lock lock;

    public Node(String name, String url) {
        this.name = name;
        this.url = url;
        this.lock = new ReentrantLock();
        this.dirty = false;
        this.cleaned = false;
    }

    public void accept(Visitor visitor) {
        parents.parallelStream().forEach(node -> node.accept(visitor));
        lock.lock();
        visitor.visit(this);
        lock.unlock();
    }

    public boolean isDirty() {
        lock.lock();
        boolean isDirty = !cleaned && (dirty || parents.stream().anyMatch(Node::isDirty));
        lock.unlock();
        return isDirty;
    }

    public void setDirty(boolean dirty) {
        this.dirty = dirty;
    }

    public boolean isCleaned() {
        return cleaned;
    }

    public void setCleaned(boolean cleaned) {
        this.cleaned = cleaned;
    }

    public List<Node> getParents() {
        return parents;
    }

    public void setParents(List<Node> parents) {
        this.parents = parents;
    }
}

File: src/main/kotlin/Node.kt
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock
import java.util.function.Consumer
import kotlin.concurrent.withLock

class Node(val name: String, val url: String) {
    var changes: Boolean = false
    var treated: Boolean = false
    var parents: List<Node> = emptyList()

    private val lock: Lock = ReentrantLock()

    fun treat(action: Consumer<Node>) {
        parents.parallelStream().forEach { it.treat(action) }
        if (mustBeTreated()) {
            lock.withLock { action.accept(this) }
        }
    }

    fun mustBeTreated(): Boolean =
        lock.withLock { return !treated && (changes || parents.any { it.treated } || parents.isEmpty()) }
}
